python-mini-projects/projects/Tic_tac_toe_with_ai/tic-tac-toe-AI.py

```diff
@@ -0,0 +1,346 @@
+        #The input above has [0].upper() in the end;
+        #So the user can enter x, X, xxxx or XXX; the input will always be taken as X;
+        #Thereby, increasing the user input window;
+
+        if p1_choice == 'X' or p1_choice == 'O':
+            #if entered value is X or O; get out of the loop; 
+            break
+        print("INVALID INPUT! Please Try Again!") 
+        #if the entered value isn't X or O, re-run the while loop;
+
+        #WHILE LOOP ENDS
+    #Assigning the value to p2 and then diplaying the values;
+    if p1_choice == 'X':
+        p2_choice = 'O'
+    elif p1_choice == 'O':
+        p2_choice = 'X'
+    
+    return (p1_choice, p2_choice)
+
+
+
+def first_player():
+    #This function will randomly decide who will go first;
+    import random
+    return random.choice((0, 1))
+
+
+def display_board(board, avail):
+    print("    " + " {} | {} | {} ".format(board[7],board[8],board[9]) + "            " + " {} | {} | {} ".format(avail[7],avail[8],avail[9]))
+    print("    " + "-----------" + "            " + "-----------")
+    print("    " + " {} | {} | {} ".format(board[4],board[5],board[6]) + "            " + " {} | {} | {} ".format(avail[4],avail[5],avail[6]))
+    print("    " + "-----------" + "            " + "-----------")
+    print("    " + " {} | {} | {} ".format(board[1],board[2],board[3]) + "            " + " {} | {} | {} ".format(avail[1],avail[2],avail[3]))
+
+
+def player_choice(board, name, choice):
+    position = 0
+    #Initialising position as 0^; so it passes through the while loop; 
+    while position not in [1,2,3,4,5,6,7,8,9] or not space_check(board, position):
+        position = int(input(f'\n{name} ({choice}), Choose your next position: (1-9) \t'))
+        
+        if position not in [1,2,3,4,5,6,7,8,9] or not space_check(board, position) or position == "": 
+            #To check whether the given position is in the set [1-9] or whether it is empty or occupied;
+            print(f"INVALID INPUT. Please Try Again!\n")   
+    print("\n")        
+    return position
+
+
+# THIS IS THEFUNCTION WHERE AI IS ADDED:
+def CompAI(board, name, choice):
+    position = 0
+    possibilities = [x for x, letter in enumerate(board) if letter == ' ' and x != 0]
+    
+    # including both X and O, since if computer will win, he will place a choice there, but if the component will win --> we have to block that move
+    for let in ['O', 'X']:
+        for i in possibilities:
+            # Creating a copy of the board everytime, placing the move and checking if it wins;
+            # Creating a copy like this  and not this boardCopy = board, since changes to boardCopy changes the original board;
+            boardCopy = board[:]
+            boardCopy[i] = let
+            if(win_check(boardCopy, let)):
+                position = i
+                return position
+
+    openCorners = [x for x in possibilities if x in [1, 3, 7, 9]]
+    
+    if len(openCorners) > 0:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To check if a container or sequence (string, list, tuple) is empty, use `if not val`. Do not compare its length using `if len(val) == 0`.

 [Learn more](https://www.python.org/dev/peps/pep-0008/#programming-recommendations#:~:text=if%20not%20seq)

Similar issue at line number 122.